lib/srv/desktop/windows_server.go

```diff
@@ -0,0 +1,330 @@
+	}
+	if cfg.AccessPoint == nil {
+		return trace.BadParameter("WindowsServiceConfig is missing AccessPoint")
+	}
+	if cfg.ConnLimiter == nil {
+		return trace.BadParameter("WindowsServiceConfig is missing ConnLimiter")
+	}
+	if err := cfg.Heartbeat.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (cfg HeartbeatConfig) CheckAndSetDefaults() error {
+	if cfg.HostUUID == "" {
+		return trace.BadParameter("HeartbeatConfig is missing HostUUID")
+	}
+	if cfg.PublicAddr == "" {
+		return trace.BadParameter("HeartbeatConfig is missing PublicAddr")
+	}
+	if cfg.OnHeartbeat == nil {
+		return trace.BadParameter("HeartbeatConfig is missing OnHeartbeat")
+	}
+	return nil
+}
+
+// NewWindowsService initializes a new WindowsService.
+//
+// To start serving connections, call Serve.
+// When done serving connections, call Close.
+func NewWindowsService(cfg WindowsServiceConfig) (*WindowsService, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctx, close := context.WithCancel(context.Background())
+	s := &WindowsService{
+		cfg: cfg,
+		middleware: &auth.Middleware{
+			AccessPoint:   cfg.AccessPoint,
+			AcceptedUsage: []string{teleport.UsageWindowsDesktopOnly},
+		},
+		closeCtx: ctx,
+		close:    close,
+	}
+
+	// TODO(awly): session recording.
+	// TODO(awly): user locking.
+
+	if err := s.startServiceHeartbeat(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := s.startStaticHostHeartbeats(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return s, nil
+}
+
+func (s *WindowsService) startServiceHeartbeat() error {
+	heartbeat, err := srv.NewHeartbeat(srv.HeartbeatConfig{
+		Context:         s.closeCtx,
+		Component:       teleport.ComponentWindowsDesktop,
+		Mode:            srv.HeartbeatModeWindowsDesktopService,
+		Announcer:       s.cfg.AccessPoint,
+		GetServerInfo:   s.getServiceHeartbeatInfo,
+		KeepAlivePeriod: apidefaults.ServerKeepAliveTTL,
+		AnnouncePeriod:  apidefaults.ServerAnnounceTTL/2 + utils.RandomDuration(apidefaults.ServerAnnounceTTL/10),
+		CheckPeriod:     defaults.HeartbeatCheckPeriod,
+		ServerTTL:       apidefaults.ServerAnnounceTTL,
+		OnHeartbeat:     s.cfg.Heartbeat.OnHeartbeat,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	go func() {
+		if err := heartbeat.Run(); err != nil {
+			s.cfg.Log.WithError(err).Error("Heartbeat ended with error")
+		}
+	}()
+	return nil
+}
+
+// startStaticHostHeartbeats spawns heartbeat routines for all static hosts in
+// this service. We use heartbeats instead of registering once at startup to
+// support expiration.
+//
+// When a WindowsService with a list of static hosts disappears, those hosts
+// should eventually get cleaned up. But they should exist as long as the
+// service itself is running.
+func (s *WindowsService) startStaticHostHeartbeats() error {
+	for _, host := range s.cfg.Heartbeat.StaticHosts {
+		heartbeat, err := srv.NewHeartbeat(srv.HeartbeatConfig{
```

Comment:
Curious why not heartbeat those as a part of WindowsService instead of doing separate heartbeats? That's what I did for DatabaseService/Database - each DatabaseService has a list of Databases (that can be statically and/or dynamically registered) that it heartbeats back. Both work, just wondering.

Reply:
As we discussed on slack, it's a slightly different model than with databases.
I chose to decouple service objects from desktop objects to make the API and enumeration/lookup of desktops easier.
The downside is with session routing - my approach assumes that any desktop can be connected to from any desktop service (except for agent mode / tunnels). If users need something more sophisticated, we'll add label-based matching of desktops to services.